### PR TITLE
Table widget

### PR DESCRIPTION
**Pull request by Hixie**
- Add x and y to the parent data, in case it's useful. In point of fact,
  I ended up not needing it.
- Make columnWidths settable rather than having a setter so that the
  code in the widgets layer is more idiomatic.
- Teach setFlatChildren about avoiding unnecessary drop/adopt loops.
- Assert that the child list length is correct more aggressively so that
  we catch the dumb bug I had more quickly next time.
- Improve the toString() to handle empty tables better.
- Resort the imports in framework.dart.
- Introduce a LocalKey for cases where you don't want to allow use of
  GlobalKey.
- Make inflateWidget and deactivateChild public for subclass use.
- Table widget.
- Cell widget.
- TableRow fake widget.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/table.dart [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+      ..rows = children.length
+      ..columnWidths = columnWidths
+      ..defaultColumnWidth = defaultColumnWidth
+      ..border = border
+      ..defaultVerticalAlignment = defaultVerticalAlignment
+      ..textBaseline = textBaseline;
+  }
+}
+
+class TableElement extends RenderObjectElement {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: _TableElement ?

File: packages/flutter/lib/src/widgets/table.dart [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+  List<_TableElementRow> _children = const<_TableElementRow>[];
+
+  bool _debugWillReattachChildren = false;
+
+  @override
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);
+    assert(!_debugWillReattachChildren);
+    _debugWillReattachChildren = true;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: Should this write be inside an assert?

File: packages/flutter/lib/src/widgets/table.dart [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    _children = widget.children.map((TableRow row) {
+      return new _TableElementRow(
+        key: row.key,
+        children: row.children.map((Widget child) => inflateWidget(child, null)).toList()
+      );
+    }).toList();
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: growable: false

File: packages/flutter/lib/src/widgets/table.dart [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        key: row.key,
+        children: updateChildren(oldChildren, row.children, detachedChildren: _detachedChildren)
+      ));
+    }
+    _debugWillReattachChildren = false;
+    updateRenderObjectChildren();
+    _detachedChildren.clear();
+  }
+
+  void updateRenderObjectChildren() {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: I probably would have made this private.

File: packages/flutter/lib/src/widgets/table.dart [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      value: (_TableElementRow row) => row.children
+    );
+    Iterator<_TableElementRow> oldUnkeyedRows = _children.where((_TableElementRow row) => row.key == null).iterator;
+    List<_TableElementRow> _newChildren = <_TableElementRow>[];
+    for (TableRow row in widget.children) {
+      List<Element> oldChildren;
+      if (row.key != null && oldKeyedRows.containsKey(row.key)) {
+        oldChildren = oldKeyedRows[row.key];
+      } else if (oldUnkeyedRows.moveNext()) {
+        oldChildren = oldUnkeyedRows.current.children;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: We probably shouldn't go down this branch if row.key != null

File: packages/flutter/lib/src/widgets/table.dart [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+      } else {
+        oldChildren = const <Element>[];
+      }
+      _newChildren.add(new _TableElementRow(
+        key: row.key,
+        children: updateChildren(oldChildren, row.children, detachedChildren: _detachedChildren)
+      ));
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: If we didn't select an old row, we need to update it against a new list of []

File: packages/flutter/lib/src/widgets/table.dart [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+  @override
+  bool detachChild(Element child) {
+    _detachedChildren.add(child);
+    deactivateChild(child);
+    return true;
+  }
+}
+
+class Cell extends ParentDataWidget<Table> {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/table.dart]: Should this be TableCell to be parallel with TableRow ?

File: packages/flutter/test/widget/table_test.dart [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter_test/flutter_test.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+import 'package:test/test.dart';
+
+void main() {
+  test('Table widget - control test', () {
```
Comment (abarth) [inline packages/flutter/test/widget/table_test.dart]: Please add a test with a StatefulBuilder and a TableRow with a key.

**Comment by Hixie** [comment]
WRITE TESTS
FIND BUGS
PTAL

**Comment by abarth** [comment]
LGTM

**Pull request merged**